src/main/java/org/xbill/DNS/NioTcpClient.java

```diff
@@ -265,8 +248,10 @@ private void processWrite() {
                     c.bind(local);
                   }
 
+                  final ChannelState channelState = new ChannelState(c);
+                  c.register(selector, SelectionKey.OP_CONNECT, channelState);
```

Comment:
AFAIR this won't work in a multithreaded scenario. A registration on the selector must be made on the selector thread or it can cause deadlocks.

Reply:
I guess you're referring to this in the [`Selector`](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/Selector.html) docs:

> A selector's key and selected-key sets are not, in general, safe for use by multiple concurrent threads. If such a thread might modify one of these sets directly then access should be controlled by synchronizing on the set itself. The iterators returned by these sets' iterator methods are fail-fast: If the set is modified after the iterator is created, in any way except by invoking the iterator's own remove method, then a ConcurrentModificationException will be thrown.

Since [`SelectableChannel.register`](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/SelectableChannel.html#register(java.nio.channels.Selector,%20int)) modifies the selectors key set, it might cause that ConcurrentModificationException to be thrown. But in general it seems that the register method itself can be called from multiple threads.

Anyhow, I think I will revert back to using the registration queue to avoid having to deal with concurrent modification exceptions on the selector thread.

Reply:
It's been a while, but I think I actually ran into deadlocks (not the ConcurrentModificationException) when registering from outside the selector thread.

Reply:
fixed